libpromises/pipes_unix.c

```diff
@@ -879,17 +881,32 @@ int cf_pclose_full_duplex_side(int fd)
 /* We are assuming that read_fd part will be always open at this point. */
 int cf_pclose_full_duplex(IOData *data)
 {
+    assert(data != NULL);
     ThreadLock(cft_count);
 
     if (CHILDREN == NULL)
     {
         ThreadUnlock(cft_count);
-        if (data->read_fd >= 0)
+        if (data->read_stream != NULL)
+        {
+            // fclose closes the underlying fd / socket,
+            // but we need the fd for handling of processes below.
+            assert(data->read_fd >= 0);
+            fclose(data->read_stream);
+        }
+        else if (data->read_fd >= 0)
         {
             close(data->read_fd);
         }
 
-        if (data->write_fd >= 0)
+        if (data->write_stream != NULL)
+        {
+            // fclose closes the underlying fd / socket,
+            // but we need the fd for handling of processes below.
+            assert(data->write_fd >= 0);
+            fclose(data->write_stream);
+        }
+        else if (data->write_fd >= 0)
```

Comment:
Should this maybe set `data->write_stream` and `data->write_fd` to `NULL` and `0` respectively?

Reply:
It could, but it didn't before, so I didn't change it. (Also, if changed, it should be `-1` for invalid fd.)